packages/malloy/src/lang/test/parse.spec.ts

```diff
@@ -303,11 +303,494 @@ describe('model statements', () => {
+          "Cannot define 's', value has unknown type"
+        );
+      });
+      test(
+        'can select different overload',
+        modelOK('query: a -> { group_by: s is concat() }')
+      );
+      test(
+        'can pass different expression types',
+        modelOK(`query: a -> {
+          group_by: f1 is sqrt(1)
+          aggregate: f2 is sqrt(count())
+          aggregate: f3 is sqrt(all(count()))
+          calculate: f4 is sqrt(lag(f1))
+          calculate: f5 is sqrt(lag(count()))
+        }`)
+      );
+      test(
+        'function return type correct',
+        modelOK(`query: a -> {
+          group_by: s is floor(1.2) + 1
+        }`)
+      );
+      test('function return type incorrect', () => {
+        expect(`query: a -> {
+            group_by: s is floor(1.2) + 'a'
+        }`).compileToFailWith(
+          "Non numeric('number,string') value with '+'",
+          "Cannot define 's', value has unknown type"
+        );
+      });
+      test(
+        'can use output value in calculate',
+        modelOK(`query: a -> {
+          group_by: x is 1
+          calculate: s is lag(x)
+        }`)
+      );
+      test('cannot use output value in group_by', () => {
+        expect(`query: a -> {
+          group_by: x is 1
+          group_by: y is x
+        }`).compileToFailWith(
+          "'x' is not defined",
+          "Cannot define 'y', value has unknown type"
+        );
+      });
+      test('lag can check that other args are constant', () => {
+        expect(`query: a -> {
+          group_by: x is 1
+          calculate: s is lag(x, x)
+        }`).compileToFailWith(
+          // TODO improve this error message
+          "Parameter 2 ('offset') of lag must be literal or constant, but received output"
+        );
+      });
+      test(
+        'lag can use constant values for other args',
+        modelOK(`query: a -> {
+          group_by: x is 1
+          calculate: s is lag(x, 2)
+        }`)
+      );
+      test('cannot name top level objects same as functions', () => {
+        expect(
+          markSource`query: ${'concat is a -> { group_by: x is 1 }'}`
+        ).compileToFailWith(
+          // TODO improve this error message
+          "'concat' is already defined, cannot redefine"
+        );
+      });
+      test(
+        '`now` is considered constant`',
+        modelOK(`query: a -> {
+          group_by: n is now
+          calculate: l is lag(n, 1, now)
+        }`)
+      );
+      // TODO this doesn't work today, we're not rigorous enough with integer
+      // subtypes. But we should probably make this typecheck properly.
+      test.skip('cannot use float in round precision', () => {
+        expect(`query: a -> {
+          group_by: x is round(1.5, 1.6)
+        }`).compileToFailWith(
+          // TODO improve this error message
+          "Parameter 2 ('precision') for round must be integer, received float"
+        );
```

Comment:
We don't do number subtypes well today. So this kind of error ends up being a SQL error instead of a compile error.